crates/bevy_animation/src/lib.rs

```diff
@@ -479,173 +513,104 @@ impl AnimationPlayer {
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-) -> bool {
-    let Some(mut current) = player_parent.map(Parent::get) else {
-        return true;
-    };
-    loop {
-        let Ok((has_player, parent)) = parents.get(current) else {
-            return true;
+        // Advance the main animation.
+        if let Some(animation_clip) = animation_clips.get(&player.animation.animation_clip) {
+            player
+                .animation
+                .update(time.delta_seconds(), animation_clip.duration);
         };
-        if has_player {
-            return false;
-        }
-        if let Some(parent) = parent {
-            current = parent.get();
-        } else {
-            return true;
+
+        // Advance transition animations.
+        for transition in &mut player.transitions {
+            // Decrease weight. But don't expire the transition yet if the time
+            // ran out; we still need to update the bones.
+            transition.current_weight = (transition.current_weight
+                - transition.weight_decline_per_sec * time.delta_seconds())
+            .max(0.0);
+
+            if let Some(animation_clip) = animation_clips.get(&transition.animation.animation_clip)
+            {
+                transition
+                    .animation
+                    .update(time.delta_seconds(), animation_clip.duration);
+            };
         }
     }
 }
 
-/// System that will play all animations, using any entity with a [`AnimationPlayer`]
-/// and a [`Handle<AnimationClip>`] as an animation root
-#[allow(clippy::too_many_arguments)]
-pub fn animation_player(
-    time: Res<Time>,
-    animations: Res<Assets<AnimationClip>>,
-    children: Query<&Children>,
-    names: Query<&Name>,
-    transforms: Query<&mut Transform>,
-    morphs: Query<&mut MorphWeights>,
-    parents: Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    mut animation_players: Query<(Entity, Option<&Parent>, &mut AnimationPlayer)>,
+/// A system that modifies animation targets (bones) according to the
+/// currently-playing animation.
+///
+pub fn animate_targets(
+    world: &mut World,
+    mut system_state: Local<
+        Option<
+            SystemState<(
+                Query<&AnimationPlayer, Without<AnimationTarget>>,
+                Query<EntityMut, With<AnimationTarget>>,
+            )>,
+        >,
+    >,
 ) {
-    animation_players
-        .par_iter_mut()
-        .for_each(|(root, maybe_parent, mut player)| {
-            update_transitions(&mut player, &time);
-            run_animation_player(
-                root,
-                player,
-                &time,
-                &animations,
-                &names,
-                &transforms,
-                &morphs,
-                maybe_parent,
-                &parents,
-                &children,
-            );
+    if system_state.is_none() {
+        *system_state = Some(SystemState::new(world));
+    }
```

Comment:
Is there a reason `system_state` can't be a `&mut SystemState<...>`?

Reply:
Didn't realize you could do that, thanks :)